libc/string_cpy.c

```diff
@@ -0,0 +1,1269 @@
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(&memStr[(BIG_NUMB / 2) - 11], expVal, sizeof(testStr) * 3 - 3);
+
+#endif
+}
+
+
+TEST(string_memccpy, one_byte)
+{
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+#ifndef __phoenix__
+
+	char testArray[MAX_ASCII_SET] = { 0 },
+		 input[MAX_ASCII_SET] = { 1 };
+	int i;
+
+	TEST_ASSERT_NOT_EQUAL_INT(0, strcmp(input, testArray));
+
+	for (i = 0; i < MAX_ASCII_SET; i++) {
+
+		/* This loop will copy only one byte in a place where the loop iterator pointing this place */
+		TEST_ASSERT_EQUAL_PTR(&testArray[i] + 1, memccpy(&testArray[i], &input[i], input[i], 1));
+	}
+
+	/* Checking if all elements were correctly copied */
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(input, testArray, MAX_ASCII_SET);
+
+#endif
+}
+
+
+TEST(string_memccpy, nothing)
+{
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+#ifndef __phoenix__
+
+	char testArray[MAX_STR_LEN] = TEST_STR1,
+		 input[MAX_STR_LEN] = { 0 },
+		 strText[] = "Lorem ipsum \0 dolor";
+
+	/* Clearing array with support array filled with 0 */
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(testArray, input, 'x', MAX_STR_LEN));
+	TEST_ASSERT_NOT_EQUAL_INT(0, strcmp(testArray, TEST_STR1));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(input, testArray, MAX_STR_LEN);
+
+
+	/* Checking no sensitivity for NULL */
+	memset(input, 0, sizeof(input));
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(input, strText, 'x', sizeof(strText)));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(strText, input, sizeof(strText));
+
+#endif
+}
+
+
+TEST(string_memccpy, big)
+{
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+#ifndef __phoenix__
+
+	char buff[BIG_NUMB] = { 0 };
+
+	/* Checking capability of handling big blocks of data */
+	TEST_ASSERT_EQUAL_INT(1, big_string());
+	memccpy(buff, long_str, '\xFF', BIG_NUMB);
+	TEST_ASSERT_EQUAL_STRING(buff, long_str);
+	TEST_ASSERT_EQUAL_CHAR(buff[0], '\1');
+	TEST_ASSERT_EQUAL_CHAR(buff[MAX_ASCII_SET / 4], 'A');
+	TEST_ASSERT_EQUAL_CHAR(buff[MAX_ASCII_SET / 2], '\x81');
+
+	/* Offset big chunk of data */
+	TEST_ASSERT_NOT_NULL(memccpy(&buff[128], long_str, '\xFF', BIG_NUMB - MAX_ASCII_SET));
+	TEST_ASSERT_EQUAL_CHAR(buff[0], '\1');
+	TEST_ASSERT_EQUAL_CHAR(buff[MAX_ASCII_SET / 4], 'A');
+	TEST_ASSERT_EQUAL_CHAR(buff[MAX_ASCII_SET / 2], '\x01');
+	TEST_ASSERT_NOT_EQUAL_CHAR(buff[MAX_ASCII_SET / 2], '\x7F');
+	TEST_ASSERT_NOT_EQUAL_CHAR(buff[MAX_ASCII_SET], '\x01');
+	TEST_ASSERT_EQUAL_CHAR(buff[(MAX_ASCII_SET / 4) + 128], 'A');
+	TEST_ASSERT_EQUAL_CHAR(buff[(MAX_ASCII_SET / 2) + 128], '\x81');
+
+#endif
+}
+
+
+/*
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////
+ */
+
+
+TEST_SETUP(string_strncpy)
+{
+}
+
+
+TEST_TEAR_DOWN(string_strncpy)
+{
+}
+
+
+TEST(string_strncpy, basic)
+{
+	int num = 12345678;
+	char payload[MAX_STR_LEN] = { 0 },
+		 buff[MAX_STR_LEN] = { 0 };
+
+	TEST_ASSERT_EQUAL_STRING(TEST_STR1, strncpy(buff, TEST_STR1, sizeof(buff)));
```

Comment:
I think it would be better to set the the number of bytes intended to copy as `n`, so in this case `sizeof(TEST_STR1) - 1` by default and create a separate case, where we will check the ability to copy only bytes before a nul teminatin character.

Reply:
After applying those changes and running the system on host-generic-pc this warning occurred:
![Screenshot from 2023-05-12 11-58-14](https://github.com/phoenix-rtos/phoenix-rtos-tests/assets/49757239/788fbd28-8b1a-4285-8aab-2070dbdf6c19)